### PR TITLE
Support for 9.4

### PR DESCRIPTION
**Pull request by edsko**
This adds support for ghc 9.4. This is currently WIP, opening a draft PR just to track the work.

### CODE REVIEW

**Comment by edsko** [comment]
@ndmitchell I think this is ready for review/merge:

* The first commit is #56 , and fixes support for `ghc` 9.2.
* The second commit changes the code that we generate to _get_ record fields, relying on `getField` instances generated by `ghc` itself. This is necessary for 9.4, because it no longer supports type-based disambiguation; something like `(name :: Company -> String)` will give an "ambiguous field" error; so now we generate `getField @"name"` instead. (Eventually type-based disambiguation will be removed for _setting_ fields also, but hopefully at that stage `ghc` itself will give us `HasField` instances.)

    **NOTE**: Technically this is a backwards incompatible change, since this requires users to enable `TypeApplications`. (We now use this new scheme for ghc versions starting at 9.0; I could change that to be 9.4, so that `TypeApplications` would only be required then -- it just provided a cleaner story to change it to be from 9.0, because then it's independent of the commits below.)
* The third commit introduces two type classes that just clean up the code a bit; the situation with location information and extensions in the ghc AST is a bit of a mess, and 9.4 was making that a lot worse; the `record-dot-preprocessor` was using `noL`, `noLA`, `noL'` and possibly others, and with 9.4 that was resulting in even more variations. Now there is just `noL`, which is a class method and is resolved correctly for all ghc versions. Similarly with `noAnn`. 
* Then the main commit just updates the code for `ghc` 9.4; nothing much of interest happening here, just the usual problems when dealing with different `ghc` versions. The annoying parts were already handled in the commits prior to this.
* Finally, the final commit is not strictly part of support for 9.4, but it adds a `cabal.project` file which, importantly, sets `write-ghc-environment-files` to `always`, which makes it possible to run `cabal test` without a global package environment. But if you don't want this commit, I can remove it again.

I have tested this (`cabal test`) against ghc 8.8, 8.10, 9.0, 9.2 and 9.4, but if you have a codebase of your own that you could run this against just to be sure, that might be good.

File: cabal.project [hunk 1]
```diff
@@ -0,0 +1,6 @@
+packages: .
```
Comment (ndmitchell) [inline cabal.project:1]: Is this necessary? Or helpful for other reasons?

**Review by ndmitchell — COMMENTED** [review]
Looks good. Can you rebase or otherwise modify the code so it triggers the CI (it had been disabled, but I've reenabled it now)

File: cabal.project (hunk 1 above)
Comment (edsko) [reply cabal.project:1]: It's necessary if you have a `cabal.project` at all (otherwise it complains that there is nothing to do).

**Comment by edsko** [comment]
@ndmitchell The CI failures were due to warnings in the test suite; ghc 9.4 does not like it if there are `LANGUAGE` or `OPTIONS_GHC`  pragmas _after_ a `module ...` declaration. 

In the process of attempting to address this however I discovered that the "installed" part of the test suite probably has been broken for quite a while. The instances that the `plugin` generates do _not_ support the test base from `Both2`; neither the setter (`r { id = ..}`) nor the getter (``id :: MyRecord -> ..`) are valid: both will report clashes with the `id` from the `Prelude`. Since this is unrelated to this PR, I've just marked `Both2` as incompatible with the plugin in `blacklist`.

To solve the `module`/pragma problem, I modified the test setup a little bit; there is now `examples/Header_in.hs` which contains the `LANGUAGE` and `OPTIONS_GHC` pragmas required, and the test suite inserts this at the right place, along with a `module` header. 

Tests now pass without warnings on my machine for ghc 8.6, 8.8, 8.10, 9.0, 9.2 and 9.4.

**Comment by edsko** [comment]
Final (hopefully) problem was `hlint`. Copying from the commit message:

    It was correctly warning about `MultiParamTypeClasses` being unused, so
    I've removed that pragma; but it was _incorrectly_ warning about
    `TypeOperators` being unused: `(~)` technically is also a type operator,
    and ghc 9.4 will start issuing a warning if you use it without the
    language pragma; so I've disabled the hint for now.

**Comment by edsko** [comment]
It seems the CI infra did not like the `cabal.project` file so removed it from this PR.

**Comment by edsko** [comment]
I'm getting some very weird interaction between this plugin and my own `large-records` plugin on 9.4; `ghc` seems to just stall entirely without using CPU. I don't yet know if the problem is here or in `large-records`; we might want to hold off on merging this PR until I do.

**Comment by edsko** [comment]
Hold off for _just_ a bit perhaps; I am looking at `record-dot-preprocessor` versus `OverloadedDot`. There might be nothing to do, I'm just not totally sure yet. I'll try to report back in a couple days?

**Comment by ndmitchell** [comment]
Sounds good - I'm in no rush. The ball is in your court.

**Comment by edsko** [comment]
No, we're good, nothing left to do. A release would be great! :)

**Comment by ndmitchell** [comment]
Released v0.2.16

**Pull request merged**